### PR TITLE
Fix metrics (especially loss) computation

### PR DESCRIPTION
**Pull request by Ranapop**
Three things are fixed with these changes:
* Fixing predicted length
  * On the inference flow the predicted length will be that configured (BOS and EOS tokens are now taken into consideration)
* Passing correct gold sequence lengths to metrics computation
  * Before these changes the lengths of the gold sequences used for metrics computation were also taking into account the BOS token, even though the sequences without BOS tokens were compared). After these changes, the lengths of the gold sequences (with EOS but not BOS) are used.
* Fixing loss mean
  * Before these changes the extra tokens were taken into consideration when computing the loss (tokens after the end of the gold sequence). Even though the tokens were masked, they were still taken into consideration when computing the mean. That means, a longer predicted sequence would make the loss smaller. Now, the loss is computed at sequence level without taking the extra tokens into consideration, then a mean is computed at batch level to compute the loss.

### CODE REVIEW

File: examples/cfq/train.py [hunk 1]
```diff
@@ -85,8 +85,10 @@ def cross_entropy_loss(logits: jnp.array, labels: jnp.array,
   labels = common_utils.onehot(labels, vocab_size)
   log_soft = nn.log_softmax(logits)
   log_sum = jnp.sum(log_soft * labels, axis=-1)
-  masked_log_sum = jnp.mean(mask_sequences(log_sum, lengths))
-  return -masked_log_sum
+  masked_log_sums = jnp.sum(mask_sequences(log_sum, lengths))
+  mean_losses = jnp.divide(masked_log_sums, lengths)
+  mean_loss = jnp.mean(mean_losses)
+  return - mean_loss
```
Comment (marcvanzee) [inline examples/cfq/train.py]:
```suggestion
  return -mean_loss
```

File: examples/cfq/train.py [hunk 2]
```diff
@@ -105,6 +107,27 @@ def pad_along_axis(array: jnp.array,
   return padded
 
 
+def compute_perfect_match_accuracy(predictions: jnp.array,
+                                   labels: jnp.array,
+                                   lengths: jnp.array) -> jnp.array:
+    """Compute perfect match accuracy.
+    
+    The accuracy is 1 only if the sequences are equal, otherwise is 0. The
```
Comment (marcvanzee) [inline examples/cfq/train.py]: This explanation makes it sound like this function will only return 0 or 1, but in fact it will return the average accuracy over all inputs. Perhaps you can clarify this?

File: examples/cfq/train.py [hunk 3]
```diff
@@ -105,6 +107,27 @@ def pad_along_axis(array: jnp.array,
   return padded
 
 
+def compute_perfect_match_accuracy(predictions: jnp.array,
```
Comment (marcvanzee) [inline examples/cfq/train.py:111]: Thanks for factoring this out!

File: examples/cfq/train_test.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+import jax.numpy as jnp
+
+
+from train import mask_sequences
+from train import cross_entropy_loss
+from train import compute_perfect_match_accuracy
+from train import compute_metrics
+
+class TrainTest(parameterized.TestCase):
```
Comment (marcvanzee) [inline examples/cfq/train_test.py:28]: More tests! Awesome! 😄

File: examples/cfq/input_pipeline.py [hunk 5]
```diff
@@ -97,7 +97,8 @@ def get_padded_shapes(self, output_len):
         constants.QUESTION_KEY: [None],
         constants.QUERY_KEY: [query_pad],
         constants.QUESTION_LEN_KEY: [],
-        constants.QUERY_LEN_KEY: []
+        constants.QUERY_LEN_KEY: [],
+        constants.QUERY_NO_BOS_LEN_KEY: []
```
Comment (marcvanzee) [inline examples/cfq/input_pipeline.py]: This seems redundant. Given that QUERY_NO_BOS_LEN_KEY is always QUERY_LEN_KEY-1, why do you have to store it separately? Can't you just compute one from the other when you need it?

**Review by marcvanzee — CHANGES_REQUESTED** [review]

File: examples/cfq/input_pipeline.py (hunk 5 above)
Comment (Ranapop) [reply examples/cfq/input_pipeline.py]: Most likely yes, but I will have to do some subtractions at batch level

File: examples/cfq/input_pipeline.py [hunk 6]
```diff
@@ -204,8 +204,9 @@ def get_bucketed_batches(self,
                                           drop_remainder=False,
                                           shuffle=True)
   batch = next(tfds.as_numpy(train_batches.skip(4)))
-  questions, queries, lengths = batch[constants.QUESTION_KEY], batch[
-      constants.QUERY_KEY], batch[constants.QUESTION_LEN_KEY]
+  questions, queries, lengths, no_bos_lengths = batch[constants.QUESTION_KEY], batch[
+      constants.QUERY_KEY], batch[constants.QUERY_LEN_KEY], batch[
+      constants.QUERY_NO_BOS_LEN_KEY]
```
Comment (marcvanzee) [inline examples/cfq/input_pipeline.py]: I think you removed this constant? Did you check the code still compiles?

**Review by marcvanzee — APPROVED** [review]
Looks good! Just make sure the code compiles and the tests are passing.
